src/main/scala/tech/sourced/gemini/Hash.scala

```diff
@@ -0,0 +1,172 @@
+    var feConfig = SparkFEClient.getConfig(session)
+    val rows = uastsDF.flatMap { row =>
+      val uastArr = row.getAs[Seq[Array[Byte]]]("uast")
+      val features = uastArr.flatMap { byteArr =>
+        val uast = Node.parseFrom(byteArr)
+        SparkFEClient.extract(uast, feConfig)
+      }
+      features.map(f => RDDFeature(RDDFeatureKey(f.name, row.getAs[String]("document")), f.weight.toLong))
+    }
+
+    rows.rdd
+  }
+
+  // TODO(max): Try to use DF here instead
+  protected def makeDocFreq(uasts: DataFrame, features: RDD[RDDFeature]): OrderedDocFreq = {
+    log.info("creating document frequencies")
+
+    val docs = uasts.select("document").distinct().count()
+
+    val df = features
+      .map(row => (row.key.token, row.key.doc))
+      .distinct()
+      .map(row => (row._1, 1))
+      .reduceByKey((a, b) => a + b)
+      .collectAsMap()
+    val tokens = df.keys.toArray[String].sorted
+
+    new OrderedDocFreq(docs.toInt, tokens, df.toMap[String, Int])
+  }
+
+  // TODO(max): Try to use DF here instead
+  protected def hashFeatures(
+                              docFreq: OrderedDocFreq,
+                              featuresRdd: RDD[RDDFeature]): RDD[RDDHash] = {
+
+    val tf = featuresRdd
+      .map(row => (row.key, row.weight))
+      .reduceByKey((a, b) => a + b)
+
+    val tfidf = tf
+      .map(row => (row._1.doc, Feature(row._1.token, row._2)))
+      .groupByKey()
+      .map { row =>
+        val doc = row._1
+        val features = row._2
+
+        RDDHash(doc, FeaturesHash.hashFeatures(docFreq, features))
+      }
+
+    tfidf
+  }
+
+  protected def saveDocFreq(docFreq: OrderedDocFreq): Unit = {
+    log.info("save document frequencies")
+    // TODO(max) replace with DB later
+    docFreq.saveToJson(Gemini.defaultDocFreqFile)
+  }
+
+  protected def saveMeta(files: DataFrame, keyspace: String, tables: Tables): Unit = {
+    log.info("save meta")
+
+    val cols = tables.metaCols
+    val renamedFiles = files
+      .select("blob_id", "repository_id", "commit_hash", "path")
+      .withColumnRenamed("blob_id", cols.sha)
+      .withColumnRenamed("repository_id", cols.repo)
+      .withColumnRenamed("commit_hash", cols.commit)
+      .withColumnRenamed("path", cols.path)
+
+    val approxFileCount = renamedFiles.rdd.countApprox(10000L)
+    log.info(s"Writing $approxFileCount files to DB")
+    renamedFiles.write
+      .mode("append")
+      .cassandraFormat(tables.meta, keyspace)
+      .save()
+  }
+
+  protected def saveHashes(rdd: RDD[RDDHash], keyspace: String, tables: Tables): Unit = {
+    log.info("save hashtables")
+
+    val cols = tables.hashtablesCols
+    rdd.flatMap(row => {
+      val RDDHash(doc, wmh) = row
+      FeaturesHash.wmhToBands(wmh).zipWithIndex.map{ case(band, i) => (doc, i, band) }
+    })
+      .toDF(cols.sha, cols.hashtable, cols.value)
+      .write
+      .mode("append")
+      .cassandraFormat(tables.hashtables, keyspace)
+      .save()
```

Comment:
A nitpick, but this formatting for method-chaining looks more consistent

```scala
    rdd
      .flatMap { row =>
        val RDDHash(doc, wmh) = row
        FeaturesHash.wmhToBands(wmh).zipWithIndex.map { case(band, i) => (doc, i, band) }
      }
      .toDF(cols.sha, cols.hashtable, cols.value)
      .write
      .mode("append")
      .cassandraFormat(tables.hashtables, keyspace)
      .save()
```

Reply:
fixed